Enable Xcode 16.1 on CI

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

Context: p91TBi-cnK-p2
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Enable Xcode 16.1 on the CI.

There are a few new unit test failures, some of which are hard to explain:

#### InAppPurchaseStoreTests

Changing `storefront` country for `SKTestSession` doesn't reset it in other tests which results in failures. Whether `SKTestSession` is cleared or recreated after each test doesn't matter. There are other users who faced this issue https://forums.developer.apple.com/forums/thread/764937. I couldn't find a solution so I'm skipping these tests

#### NoWooErrorViewModelTests

One unit test case relies on `popToRootViewController` which no longer works in iOS 18 + Xcode 18 unit tests. I found someone else reported the issue https://forums.developer.apple.com/forums/thread/759640  and Apple asked to report it. For now, I removed this part of the test.

#### Order UI Tests

New order creation UI tests fail due to changes in default modal sizes on iPadOS 18 (p91TBi-c9Y-p2). There's no official communication on this but it looks like the default size changed to `.form`. In this particular failed test, we open the 'Add Customer address' modal and expect all fields to be visible. Since this view displays more than a dozen fields, I think it makes sense to explicitly display it in full screen. 

#### Payment Tests - Learn More link

The learn More link in the Payment view is made from the attributed text. UI tests can find and `tap()` on this element but nothing happens. I tried different approaches and only `click()` helped to solve the issue without rebuilding the internals of the label. However, `click()` only works on iPad therefore I'm skipping this test on the iPhone.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

The only user-facing change in this PR is explicitly presenting customer address details in full-screen:

https://github.com/user-attachments/assets/fdbdaa8e-07da-4655-8741-8770bc7c0c39

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Enable Xcode 16 on CI

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Disable failing StoreKit tests

iangmaia commented:
Hi @staskus! FYI, I have [this PR](https://github.com/woocommerce/woocommerce-ios/pull/14207) also updating the Xcode image on CI to `16.1-rc1` -- though my focus there is not on getting the project to compile and work correctly with Xcode 16, which I guess is you focus for this one.
I see the UI tests are failing on both PRs, so perhaps more changes are needed?

I'll rebase and stack my PR on top of yours, so that this one can focus on getting 16 fully working and #14207 will focus on the other changes.

staskus commented:
@iangmaia, yes it looks like more changes are needed. I need to inspect tests one by one to understand why they fail. I will focus on tests succeeding in this PR.

staskus committed:
Present Customer details view full screen on iPad to fit all the content

On iOS 18 default sheet size reduced. Making the Customers view to be displayed full screen to fit more content.

staskus committed:
Remove unit tests tests that rely on popToRootViewController

popToRootViewController animated doesn't on Xcode 16 + iOS 18 unit tests for unknown reason https://forums.developer.apple.com/forums/thread/759640

staskus committed:
Change tap to click for attributed link to be hittable on iOS 18 UI tests

staskus committed:
Explicitly declare Calendar type in OrderStatsV4Interval for consistent results

staskus committed:
Update NoWooErrorViewModelTests.swift

staskus committed:
Do not run test_load_learn_more_link on iPhone

Skipping the test on the iPhone. Calling tap() on a link within attributed text no longer works in UI tests on Xcode 16. Using click() instead, which is only supported on iPad.

staskus commented:
I am tagging a few developers to raise awareness and to get the review. I'm not too happy that some tests need to be skipped. Maybe for some cases, you would have some other ideas we could try. 🤔

staskus committed:
Xcode 16.1

staskus committed:
Update pipeline.yml

staskus commented on .xcode-version:
```diff
@@ -1 +1 @@
-~> 15.4
+~> 16.1-rc.1
```
```suggestion
~> 16.1
```

staskus replied:
paaHJt-7kP-p2

mokagio commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/CustomerSection/OrderCustomerSection.swift:
```diff
@@ -39,7 +39,7 @@ struct OrderCustomerSection: View {
             }
         }
-        .sheet(isPresented: $viewModel.showsAddressForm) {
+        .fullScreenCover(isPresented: $viewModel.showsAddressForm) {
```
@staskus  This is the one for which you wrote

> Since this view displays more than a dozen fields, I think it makes sense to explicitly display it in full screen [to make all fields visible].

If that's the case, sounds good to me.

mokagio commented on WooCommerce/WooCommerceTests/Authentication/NoWooErrorViewModelTests.swift:
```diff
@@ -111,18 +111,13 @@ final class NoWooErrorViewModelTests: XCTestCase {
         // Then
         XCTAssertFalse(stores.isAuthenticated)
-        XCTAssertEqual(navigationController.viewControllers.count, 1)
-        XCTAssertFalse(navigationController.topViewController is ULErrorViewController)
```
> One unit test case relies on popToRootViewController which no longer works in iOS 18 + Xcode 18 unit tests. I found someone else reported the issue [forums.developer.apple.com/forums/thread/759640](https://forums.developer.apple.com/forums/thread/759640) and Apple asked to report it. For now, I removed this part of the test.

Sounds good. Testing the popping of the nav stack was handy but also a bit "smelly". It would be good if `NoWooErrorViewModel` was agnostic of its presentation mode so that it's up to whichever component presents it to handle how to dismiss the screen the view model drives. But I think a lot of this will have to be reworked when, inevitably, this will be updated to Swift UI.

mokagio commented on WooCommerce/WooCommerceTests/UnitTests.xctestplan:
```diff
@@ -31,6 +31,8 @@
       "skippedTests" : [
         "CardReaderConnectionControllerTests",
+        "InAppPurchaseStoreTests\/test_iap_supported_in_canada()",
+        "InAppPurchaseStoreTests\/test_load_products_fails_if_iap_unsupported()",
```
Have you considered using `XCTSkip` to skip the tests? 

Not a big deal, but I personally prefer that approach because:

- It's closer to the actual test being skipped, which makes it easier to notice if you are looking at the test code, something that happens more often than looking at the test plan GUI view.
- It logs information about the test having been skipped in the console and test report, maintaining awareness about it

mokagio replied:
A more complex alternative would be to use `XCTExpectFailure`, more info pbzQyC-423-p2#comment-4208

The value there would be that if something changes in a new version of Xcode so that the tests can now pass, the fact that they pass would trigger a failure (expected to fail, but passed) allowing us to updated the tests accordingly.

staskus replied:
Thanks. Yes, it makes sense. Having it in the code leaves it less obscure. Checking skipped tests within a test plan is likely not a part of developers' workflow.

mokagio commented on WooCommerce/WooCommerceTests/Yosemite/InAppPurchaseStoreTests.swift:
```diff
@@ -62,6 +62,10 @@ final class InAppPurchaseStoreTests: XCTestCase {
+    /// Skipped
+    /// Setting storefront value in Xcode 16 in one test no longer resets in other tests, even when resetToDefaultState is called
+    /// Reason unknown
+    /// https://forums.developer.apple.com/forums/thread/764937
```
Thank you for adding the link.

Following up on my suggestion above regarding `XCTSkip`, one could use a similar approach as 

https://github.com/woocommerce/woocommerce-ios/blob/38a1190bd9e74901208b6c3d92e80e0fa5711722/WooCommerce/WooCommerceTests/POS/ViewModels/TotalsViewModelTests.swift#L215

to print this info in the test report.

Maybe also DRY via 

```swift
func skipBecauseOfStoreFrontXcode16Issue() throws {
    try XCTSkip(true, "...")
}

mokagio commented on WooCommerce/WooCommerceUITests/Tests/PaymentsTests.swift:
```diff
@@ -22,7 +22,17 @@ final class PaymentsTests: XCTestCase {
+                Using click() instead, which is only supported on iPad.
+                """
+            )
+        }
```
What do you think of moving this check in the `click` method os that every caller gets it ? 

I realize there might not be other callers ever, but it seems structurally better.

staskus replied:
Yes, it makes sense. iPhone test will test at least part of the flow.

mokagio reviewed (approved):
Looking good @staskus . Apologies for the late feedback. Notice how all my comments are nitpicks.

staskus committed:
Skip failing Xcode 16 IAP tests with XCTSkipIf

staskus committed:
Remove Xcode 16 SDK check

staskus committed:
Move XCTSkipIf to clickLearnMoreIPPLink

staskus committed:
Merge branch 'trunk' into task/enable-xcode-16

staskus committed:
Update RELEASE-NOTES.txt

staskus committed:
Update .xcode-version to 16.1

staskus committed:
Update InAppPurchaseStoreTests.swift

staskus committed:
Skip clickLearnMoreIPPLink on iPhone

staskus committed:
Use current timezone when calculating a range between dates in tests
